### PR TITLE
compile in shaders to webrender instead of requiring a res directory

### PR DESCRIPTION
**Pull request by vvuk**
Fixes issue #352.

res directory can be used to override.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/563)

### CODE REVIEW

**Comment by samlh** [comment]
This seems to be missing build.rs

**Comment by vvuk** [comment]
It was indeed :x

File: webrender/src/device.rs [hunk 1]
```diff
@@ -67,6 +68,29 @@ pub enum VertexFormat {
...
+    if let Some(ref base) = *base_path {
+        let shader_path = base.join(shader_name).with_extension("glsl");
+        if shader_path.exists() {
+            let mut source = String::new();
+            File::open(&shader_path).unwrap().read_to_string(&mut source).unwrap();
+            return Some(source);
+        }
+    }
+
+    if let Some(src) = shader_source::SHADERS.get(shader_name) {
```
Comment (glennw) [inline webrender/src/device.rs]: Use and_then() to remove the if/else here.

File: webrender/src/device.rs [hunk 2]
```diff
@@ -67,6 +68,29 @@ pub enum VertexFormat {
...
+        Some((*src).to_owned())
+    } else {
+        None
+    }
+}
+
+fn get_shader_source(shader_name: &str, base_path: &Option<PathBuf>) -> String {
+    get_optional_shader_source(shader_name, base_path).unwrap_or_else(|| {
```
Comment (glennw) [inline webrender/src/device.rs]: expect() instead of unwrap_or_else()

**Comment by glennw** [comment]
@vvuk Oh, travis is failing too - might need to update the cargo.lock for sample/replay.

**Pull request merged**